### PR TITLE
Add .editorconfig

### LINKED ISSUE #5515: Define code-style guide and add .editorconfig to the repo
I don't think there is a defined set of code-style rules. We should add one and start enforcing it on the new code especially considering that we are productizing Azure.Base.

We might want to copy https://github.com/dotnet/corefx/blob/master/.editorconfig or some other established repo style.

cc @mikeharder @weshaggard

### PR DESCRIPTION
**Pull request by pakrym**
Fixes: https://github.com/Azure/azure-sdk-for-net/issues/5515

I took C# part of corefx and only added `dotnet_style_require_accessibility_modifiers`.

I think the most contentious part here is the var usage `only use var when it's obvious what the variable type is` everything else it pretty much what we do anyway.

These settings are supported by VS (it would show a suggestion and a fix for most of these rules) and some are supported by `dotnet format` (https://github.com/dotnet/format) that can run from command line.

I don't expect we are going to reformat existing code but do expect this style to be applied to all new/changed code and enforce it in code reviews.

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+csharp_style_var_for_built_in_types = false:none
+csharp_style_var_when_type_is_apparent = false:none
+csharp_style_var_elsewhere = false:suggestion
+
+# use language keywords instead of BCL types
+dotnet_style_predefined_type_for_locals_parameters_members = true:suggestion
+dotnet_style_predefined_type_for_member_access = true:suggestion
+
+# name all constant fields using PascalCase
+dotnet_naming_rule.constant_fields_should_be_pascal_case.severity = suggestion
```
Comment (mikeharder) [inline .editorconfig]: Nit: fix spacing before equals sign

File: .editorconfig [hunk 2]
```diff
@@ -0,0 +1,133 @@
+# editorconfig.org
```
Comment (mikeharder) [inline .editorconfig]: I would like to somehow indicate which settings are different from the VS defaults, maybe append a comment to the end of the line like `# vs-default: <default value>`?

Comment (mikeharder) [reply .editorconfig]: So far, most of the settings appear to be the VS defaults.

File: .editorconfig [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+# Indentation preferences
+csharp_indent_block_contents = true
+csharp_indent_braces = false
+csharp_indent_case_contents = true
+csharp_indent_switch_labels = true
+csharp_indent_labels = one_less_than_current
+
+# avoid this. unless absolutely necessary
+dotnet_style_qualification_for_field = false:suggestion
```
Comment (mikeharder) [inline .editorconfig]: [next 4 settings] `# vs-default: false:none`

File: .editorconfig [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+dotnet_style_qualification_for_property = false:suggestion
+dotnet_style_qualification_for_method = false:suggestion
+dotnet_style_qualification_for_event = false:suggestion
+
+# only use var when it's obvious what the variable type is
+csharp_style_var_for_built_in_types = false:none
```
Comment (mikeharder) [inline .editorconfig]: [next 3 settings] I personally like using `var` whenever possible, so my preference would be to use either `true:none` (the VS default) or `true:suggestion`.  We may want to omit these settings for now, until we can reach a consensus.

File: .editorconfig [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+csharp_style_var_when_type_is_apparent = false:none
+csharp_style_var_elsewhere = false:suggestion
+
+# use language keywords instead of BCL types
+dotnet_style_predefined_type_for_locals_parameters_members = true:suggestion
```
Comment (mikeharder) [inline .editorconfig]: [next 2 settings] `# vs-default true:none`

File: .editorconfig [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+dotnet_naming_rule.constant_fields_should_be_pascal_case.symbols  = constant_fields
+dotnet_naming_rule.constant_fields_should_be_pascal_case.style    = pascal_case_style
+
+dotnet_naming_symbols.constant_fields.applicable_kinds   = field
+dotnet_naming_symbols.constant_fields.required_modifiers = const
+
+dotnet_naming_style.pascal_case_style.capitalization = pascal_case
+
+# static fields should have s_ prefix
+dotnet_naming_rule.static_fields_should_have_prefix.severity = suggestion
```
Comment (mikeharder) [inline .editorconfig]: Nit: Fix spacing

File: .editorconfig [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+dotnet_naming_symbols.private_internal_fields.applicable_kinds = field
+dotnet_naming_symbols.private_internal_fields.applicable_accessibilities = private, internal
+
+dotnet_naming_style.camel_case_underscore_style.required_prefix = _
+dotnet_naming_style.camel_case_underscore_style.capitalization = camel_case
+
+# Code style defaults
+dotnet_sort_system_directives_first = true
+csharp_preserve_single_line_blocks = true
+csharp_preserve_single_line_statements = false
```
Comment (mikeharder) [inline .editorconfig]: `# vs-default: true`

File: .editorconfig [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+csharp_space_between_method_call_name_and_opening_parenthesis = false
+csharp_space_between_method_call_parameter_list_parentheses = false
+csharp_space_between_method_declaration_empty_parameter_list_parentheses = false
+csharp_space_between_method_declaration_name_and_open_parenthesis = false
+csharp_space_between_method_declaration_parameter_list_parentheses = false
+csharp_space_between_parentheses = false
+csharp_space_between_square_brackets = false
+
+# Require accessibility modifiers
+dotnet_style_require_accessibility_modifiers = for_non_interface_members:suggestion
```
Comment (mikeharder) [inline .editorconfig]: Nit: trailing newline (which this very file should prevent in the future)

File: .editorconfig [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+csharp_new_line_between_query_expression_clauses = true
+
+# Indentation preferences
+csharp_indent_block_contents = true
+csharp_indent_braces = false
+csharp_indent_case_contents = true
+csharp_indent_switch_labels = true
+csharp_indent_labels = one_less_than_current
+
+# avoid this. unless absolutely necessary
```
Comment (jsquire) [inline .editorconfig]: I'd love to open this (no pun intended) for wider discussion.   Personally, I'm a fan of explicit scope using `this.XX` and `TypeName.XX`

**Review by mikeharder — DISMISSED** [review]

File: .editorconfig [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+dotnet_style_qualification_for_field = false:suggestion
+dotnet_style_qualification_for_property = false:suggestion
+dotnet_style_qualification_for_method = false:suggestion
+dotnet_style_qualification_for_event = false:suggestion
+
+# only use var when it's obvious what the variable type is
```
Comment (jsquire) [inline .editorconfig]: Type inference is a good thing; I'm in the "var all the things" camp.  There's plenty of semantic information around typing in modern IDEs.

File: .editorconfig [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+
+# static fields should have s_ prefix
+dotnet_naming_rule.static_fields_should_have_prefix.severity = suggestion
+dotnet_naming_rule.static_fields_should_have_prefix.symbols  = static_fields
+dotnet_naming_rule.static_fields_should_have_prefix.style    = static_prefix_style
+
+dotnet_naming_symbols.static_fields.applicable_kinds   = field
+dotnet_naming_symbols.static_fields.required_modifiers = static
+
+dotnet_naming_style.static_prefix_style.required_prefix = s_
```
Comment (mikeharder) [inline .editorconfig]: Personally, I prefer using `_` for both static and instance fields.  `s_` is a bit too close to Hungarian Notation for my tastes.

File: .editorconfig [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+dotnet_naming_rule.constant_fields_should_be_pascal_case.severity = suggestion
+dotnet_naming_rule.constant_fields_should_be_pascal_case.symbols  = constant_fields
+dotnet_naming_rule.constant_fields_should_be_pascal_case.style    = pascal_case_style
+
+dotnet_naming_symbols.constant_fields.applicable_kinds   = field
+dotnet_naming_symbols.constant_fields.required_modifiers = const
+
+dotnet_naming_style.pascal_case_style.capitalization = pascal_case
+
+# static fields should have s_ prefix
```
Comment (jsquire) [inline .editorconfig]: I'd like to open this for discussion.   In my opinion, hungarian variants are a legacy that should go away.   There is plenty of semantic information available in modern IDE that obviate the need to decorate variables.   

IMHO, Names should hold semantic meaning about what is in them, not type or scope information.

File: .editorconfig [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+
+# internal and private fields should be _camelCase
+dotnet_naming_rule.camel_case_for_private_internal_fields.severity = suggestion
+dotnet_naming_rule.camel_case_for_private_internal_fields.symbols  = private_internal_fields
+dotnet_naming_rule.camel_case_for_private_internal_fields.style    = camel_case_underscore_style
+
+dotnet_naming_symbols.private_internal_fields.applicable_kinds = field
+dotnet_naming_symbols.private_internal_fields.applicable_accessibilities = private, internal
+
+dotnet_naming_style.camel_case_underscore_style.required_prefix = _
```
Comment (jsquire) [inline .editorconfig]: I'd like to open this for discussion. In my opinion, hungarian variants are a legacy that should go away. There is plenty of semantic information available in modern IDE that obviate the need to decorate variables.   The exception that I'd make here is the backing store for a property to denote that it should only be accessed through it's owner.

IMHO, Names should hold semantic meaning about what is in them, not type or scope information.

File: .editorconfig (hunk 10 above)
Comment (weshaggard) [reply .editorconfig]: Type inference may be a good thing when you are writing the code and perhaps acceptable if you are working with the code in a modern IDE that allows you to easy jump around the code and see the actual types. However personally I do not believe type inference is a good thing for the long-term maintenance and readability of the code when you are viewing the code on github or just looking at it as normal text. From my experiences I would never recommend using var unless it is clear what the type of the variable is by simply reading the text of the code. There is a lot of information to be had by knowing the type of the variables to enable you to dig deeper if necessary but not knowing the type of a variable you end up having to dig backwards as opposed to moving forward when reading through the code, it adds more work and takes longer to understand the code. Having the types helps with self-documenting the code which helps with understanding the code.

File: .editorconfig (hunk 12 above)
Comment (weshaggard) [reply .editorconfig]: Similar to my argument for not using var I believe using `_` or `s_` help you understand the scope of the variable while just reading the code and without needing to jump around in the code file to understand that scope. While this has roots in Hungarian notation it certainly isn't full Hungarian notation which I agree isn't necessary. I do however like being able to understand the scope of a variable at a glance.

File: .editorconfig (hunk 11 above)
Comment (weshaggard) [reply .editorconfig]: For me personally I make different assumptions about variables that are member vs static and so knowing that information at a glance is generally useful.

File: .editorconfig (hunk 2 above)
Comment (weshaggard) [reply .editorconfig]: Should we simply leave out the VS defaults? At least for the VS specific settings for others that might be shared with other editors with different defaults they are still useful.

File: .editorconfig (hunk 9 above)
Comment (weshaggard) [reply .editorconfig]: Given the other rules about using `_` for members there isn't much of a need for having `this.` because it is obvious from the name things are members. If we decide against using `_`, which I see you also suggest, then I agree we should consider using `this.`.

File: .editorconfig (hunk 2 above)
Comment (mikeharder) [reply .editorconfig]: I considered this, but I think it's best to include them since they may be respected by other tools (e.g. dotnet-format, VS Code), either now or in the future.

File: .editorconfig (hunk 9 above)
Comment (jsquire) [reply .editorconfig]: ... That pretty much sums up my preferences.    Steer clear of psudo-Hungarian prefixes in favor of the constructs in the language to clarify the scope.

File: .editorconfig (hunk 10 above)
Comment (jsquire) [reply .editorconfig]: I agree with this statement at a conceptual level: code that is clear, expressive, and understandable when being read is important and a key goal.

That said, I respectfully disagree that the underlying types are deserving of being a focal point.  Respectfully, if the intent and context of the code isn't clear from member names, code structure, and associated comments, I'd argue that the author has already failed in the fundamentals; the types used aren't going to bridge that gap and rescue things.   I'd double down on that stance and say that it applies even more so to non-editor scenarios, like reading on GitHub.

In my opinion, type awareness should be, first and foremost, a tooling and compiler concern, used in service of compile-time safety and providing of a rich developer experience.  While I'll certainly agree that it is helpful when exploring code to be able inspect and explore types, I believe that needing to have a deep awareness of the types in a given block of code to reason about it is often a sign of poor design, side-effects, and/or a misuse of OO principles.   

In rebuttal, I'd argue that a reader of the code shouldn't need to be jumping around to understand a given block of code.  It should be clear and digestible in its own context.  Though, I would agree that there should be an easily discoverable and logical path to uncover types within that context to peel back the layers of the onion and dig deeper.   

In summary, I agree that the types should be clear, though I believe that should happen naturally as a output of well written and well structured code.  I'm of the opinion that if you need to explicitly call out types that there are most likely underlying problems with the code or design.

File: .editorconfig (hunk 12 above)
Comment (jsquire) [reply .editorconfig]: So would using the language constructs of `this.` and `TypeName.` without adopting a prefix system.    I'm in favor of providing the explicit semantics, just in a different way.  :)

**Comment by pakrym** [comment]
Another thing to consider is that we shouldn't diverge too far from other opensource repos (dotnet/aspnet) to avoid forcing external contributor to learn many code styles.

**Comment by AlexGhiondea** [comment]
@pakrym thanks for starting the conversation about this! As you can see, there are multiple valid points being raised :). 

In addition to what has already been discussed here, we need to make sure that the other code owners in this repo, who are not on our team, are ok with enabling *any* of these rules for their code.

Is there a way to scope these changes to just a set of directories?

**Comment by pakrym** [comment]
> Is there a way to scope these changes to just a set of directories?

Yes, we can put a copy of our `.editorconfig` into subdirectories we own. It would require a bit more effort to keep track of all files when making changes but I don't think it would happen too often.

**Comment by AlexGhiondea** [comment]
But that would require that each new library we add will have to have this file added into it.

I think we should start with a set of guidelines for coding (similar to those coming from the CoreFx team) and hold off from automatically enforcing this for now.

**Comment by pakrym** [comment]
> But that would require that each new library we add will have to have this file added into it.

I would argue that adding a file is easier than manually enforcing the style guide but I'm open to any solution that results in agreed on coding style.

**Comment by mikeharder** [comment]
Once the desired developer experience is defined, the engineering system team can help implement it.  Specifically:

1. We already have a tool to detect missing or insufficient `Readme.md` files in each package.  I suspect the tool could be easily extended to also cover `.editorconfig`.
2. We could create a tool to manage all the `.editorconfig` files (i.e. run a single command to update all files).

**Comment by AlexGhiondea** [comment]
I had a chat with @KrzysztofCwalina about this. We decided to use whatever CoreFx uses.

As for the location of the file -- let's put it in the folders that represent Track 2 client libraries. We can figure out how we are going to manage that similar to what @mikeharder was saying.

**Comment by pakrym** [comment]
I've annotated values with vs defaults that I was able to find. I've also copied `.editorconfig` into `Azure.Base` and `Azure.ApplicationModel.Configuration`

**Review by AlexGhiondea — APPROVED** [review]

**Review by weshaggard — APPROVED** [review]
LGTM

Should we file a follow-up issue for how to keep these in-sync or perhaps get agreement from the management team that they will share these and we can put it in the root?

**Comment by pakrym** [comment]
Filed https://github.com/Azure/azure-sdk-for-net/issues/5694
